keps/sig-node/4742-node-topology-downward-api/README.md

```diff
@@ -0,0 +1,409 @@
+This approach eliminates the need for workarounds, such as custom init containers with elevated privileges,
+promoting a more secure and consistent solution.
+
+## Motivation
+
+Topology awareness is crucial for a growing number of applications and workloads on Kubernetes.
+Knowing the Pods's location within the cluster's topology allows for significant performance
+optimizations and improved resilience. End-user feedback has highlighted the following key use cases:
+* High-bandwidth AI/ML workloads, especially those employing distributed training or inference,
+  demonstrate a substantial need for topology awareness. The performance of these workloads is
+  highly sensitive to communication latency between GPUs. By preferentially scheduling Pods
+  requiring GPU-to-GPU communication within the same zone or rack, training and inference
+  times can be dramatically reduced.  Importantly, while Kubernetes can facilitate this topology-aware
+  placement, the actual orchestration of optimized GPU-to-GPU communication often relies on topology awareness
+  within the machine learning framework itself (e.g., Ray, PyTorch Distributed. etc).  Kubernetes provides the
+  necessary foundation for these frameworks to operate efficiently, but the frameworks are ultimately responsible
+  for leveraging that foundation.
+* CNI plugins can leverage topology information to establish more efficient network paths,
+  reducing latency and increasing throughput. For instance, a CNI could prioritize connections
+  within the same availability zone or rack.
+* Distributed stateful applications, such as sharded databases, can use topology information to improve fault tolerance.
+  By spreading replicas and traffic across different failure domains (e.g., zones, racks), these applications can achieve
+  higher availability and resilience to failures in any one topology.
+
+Today, applications obtain topology information by using an init container to query the Kubernetes API for the underlying node,
+extracting the necessary data from the Node resource. As more workloads and use cases benefit from topology awareness,
+we aim to simplify access to this information for Pods via the Downward API.
+
+### Goals
+
+* Values from Node labels `topology.k8s.io/zone`, `topology.k8s.io/region` and `kubernetes.io/hostname` are made
+  available via downward API
+* Additional node labels can be made available via downward API using admission webhooks that mutate `pods/binding`.
+
+### Non-Goals
+
+* Exposing non-standard node labels by default
+* Enhnacements to topology-aware scheduling
+* Changes to standard topology labels in Kubernetes
+* Changes to downward API
+
+## Proposal
+
+Retreival of topology information will be achieved through the existing mechanism of passing down Pod labels using Downward API.
+A new admission plugin, `PodTopologyLabels`, will be added to kube-apiserver. When enabled, it will mutate the `pods/binding` subresource,
```

Comment:
> When enabled

Just to clarify, it will be enabled whenever the feature gate is turned on, correct? So on all new clusters in a year or so, all Pods will always have their topology labels, correct?

Reply:
in other words, there is nothing that needs configuring. And any Pod, even those that do not use downward API will have their labels populated

Reply:
That's correct, but I think before Beta, we should have another debate about whether `PodTopologyLabels` is enabled-by-default built-in admission. 

My biggest gripe about this proposal is the idea of cluttering Pod labels by forcing the topology labels on all Pods.

Reply:
That's correct, but I think before Beta, we should have another debate about whether `PodTopologyLabels` is enabled-by-default built-in admission. 

My biggest gripe about this proposal is the idea of cluttering Pod labels by forcing the topology labels on all Pods. cc @thockin

Reply:
I hear the gripe - let's push ahead and decide if we need something like a way to "request" such labels be populated (e.g. a field in spec)

Reply:
The idea that all Pods are attributed with the topology info is very attractive for me. So I agree we should continue with this and listen for the feedback.

Reply:
Sounds like there are two separate (yet related) questions:
 - Whether to enable `PodTopologyLabels` the admission by default
 - Whether to allow certain pods to opt in or out this feature.

One potential issue I see is that if we do this by default for all pods, it might cause problems for users/pods that do not want this feature for whatever reason. And we're unlikely to get feedback from them until this is made the default.

Reply:
I think this is somewhat mitigated by the default admission plugin only copying across the three labels described in the 'goals' section (`topology.k8s.io/zone`, `topology.k8s.io/region` and `kubernetes.io/hostname`).

So the only component that could cause this to 'bloat' is the scheduler or any administrator-configured admission plugins, which seems reasonable.